x/batch/keeper/delegation.go

```diff
@@ -0,0 +1,130 @@
+package keeper
+
+import (
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/distribution/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+// increment the reference count for a historical rewards value
+func (k Keeper) incrementReferenceCount(ctx sdk.Context, valAddr sdk.ValAddress, period uint64) {
+	historical := k.distrKeeper.GetValidatorHistoricalRewards(ctx, valAddr, period)
+	if historical.ReferenceCount > 2 {
+		panic("reference count should never exceed 2")
+	}
+	historical.ReferenceCount++
+	k.distrKeeper.SetValidatorHistoricalRewards(ctx, valAddr, period, historical)
+}
+
+// decrement the reference count for a historical rewards value, and delete if zero references remain
+func (k Keeper) decrementReferenceCount(ctx sdk.Context, valAddr sdk.ValAddress, period uint64) {
+	historical := k.distrKeeper.GetValidatorHistoricalRewards(ctx, valAddr, period)
+	if historical.ReferenceCount == 0 {
+		panic("cannot set negative reference count")
+	}
+	historical.ReferenceCount--
+	if historical.ReferenceCount == 0 {
+		k.distrKeeper.DeleteValidatorHistoricalReward(ctx, valAddr, period)
+	} else {
+		k.distrKeeper.SetValidatorHistoricalRewards(ctx, valAddr, period, historical)
+	}
+}
+
+// initialize starting info for a new delegation
+func (k Keeper) initializeDelegation(ctx sdk.Context, val sdk.ValAddress, del sdk.AccAddress) {
+	// period has already been incremented - we want to store the period ended by this delegation action
+	previousPeriod := k.distrKeeper.GetValidatorCurrentRewards(ctx, val).Period - 1
+
+	// increment reference count for the period we're going to track
+	k.incrementReferenceCount(ctx, val, previousPeriod)
+
+	validator := k.stakingKeeper.Validator(ctx, val)
+	delegation := k.stakingKeeper.Delegation(ctx, del, val)
+
+	// calculate delegation stake in tokens
+	// we don't store directly, so multiply delegation shares * (tokens per share)
+	// note: necessary to truncate so we don't allow withdrawing more rewards than owed
+	stake := validator.TokensFromSharesTruncated(delegation.GetShares())
+	k.distrKeeper.SetDelegatorStartingInfo(ctx, val, del, types.NewDelegatorStartingInfo(previousPeriod, stake, uint64(ctx.BlockHeight())))
+}
+
+func (k Keeper) withdrawAllDelegationRewards(ctx sdk.Context, delAddr sdk.AccAddress) (sdk.Coins, error) {
+	rewardsTotal := sdk.Coins{}
+	remainderTotal := sdk.DecCoins{}
+	k.stakingKeeper.IterateDelegations(ctx, delAddr, func(_ int64, del stakingtypes.DelegationI) (stop bool) {
+		valAddr := del.GetValidatorAddr()
+		val := k.stakingKeeper.Validator(ctx, valAddr)
+
+		// check existence of delegator starting info
+		if !k.distrKeeper.HasDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr()) {
+			return false
+		}
+
+		// end current period and calculate rewards
```

Comment:
similarly here, let's reference where from x/distribution this code comes from, for when it changes upstream

Reply:
ok